### ISSUE TITLE
Changing item set visibility causes unexpected behavior

### ISSUE THREAD

**Issue by jimsafley** [omeka/omeka-s#174]
Condition:

1. user1 creates **ItemSet1**, makes it public
1. user1 creates **ItemSet2**, makes it public
1. user2 creates **Item**, assigns it to **ItemSet1** and **ItemSet2**
1. user1 makes **ItemSet2** private

Bugs:
  * user2 sees **ItemSet2** listed in **Item** show sidebar
  * user2 will unassign **ItemSet2** from **Item** when editing (because **ItemSet2** is not in the request)

The issue here is that we're getting item sets directly from the item entity (via Doctrine's entity collections), which does not filter results based on visibility. We could use [collection filters](http://doctrine-orm.readthedocs.org/en/latest/reference/working-with-associations.html#filtering-collections) or possibly [SQL filters](http://doctrine-orm.readthedocs.org/en/latest/reference/filters.html), but we're already using the `API_SEARCH_QUERY` and `API_FIND_QUERY` events to filter API query results, so we may find a solution there.

I've successfully tested *collection filters*, but that loads the entire collection before filtering. *SQL filters* seem promising given that one filter works across all queries (and a potential replacement for the API query events above). Using an *API search subrequest* would filter the results automatically, but it would require rewriting how item sets are hydrated from within an item (we leverage entity collections liberally during hydration).

**Comment by zerocrates** [comment]
I think the collection Criteria support can do some filtering lazily:

http://docs.doctrine-project.org/en/latest/changelog/migration_2_5.html#criteria-api-support-for-extra-lazy

not sure if that's applicable here but it seems like it might be

**Comment by jimsafley** [comment]
That does help. Another problem with collection filters is that we would likely need to implement Criteria in more than one place, unlike the API events and (possibly) the SQL filter. It would be nice to filter for resource visibility all at once, like we do for API requests.

**Comment by jimsafley** [comment]
Regrettably, we can't use Doctrine's Criteria feature (collection filters) because it's incompatible with our Resource -> Item Set inheritance scheme. I'm getting the following error whenever I apply criteria to an item set collection:

```
Fatal error:  Uncaught exception 'InvalidArgumentException' with message
'ResultSetMapping builder does not currently support your inheritance scheme.'
in /var/www/html/omeka-s/vendor/doctrine/orm/lib/Doctrine/ORM/Query/ResultSetMappingBuilder.php:146
```

**Comment by jimsafley** [comment]
This and countless other bugs fixed in 3cd8455ce26bc87caadfa6eb4efa557cb5541acd (let's hope)
